### ISSUE TITLE
Ensure our generated code meets our stylecop bar

### ISSUE THREAD

**Issue by crutkas** [Microsoft/WindowsTemplateStudio#461]
We want to be sure the template code also meets our style cop quality bar.  This should not include a style cop integration for the final template, but should happen during our testing and know that we are producing code that will not throw errors.

**Comment by ralarcon** [comment]
@crutkas @mrlacey We have in mind a solution approach for this. We can implement a feature which includes the style cop nuget reference + the global supressions we consider + the build configuration to enforce it.

The feature could be hidden if we want only add it in our build process or even, provide it publicly to allow users to take advance of it...

What dou you think?

**Comment by crutkas** [comment]
we can add it as an option to expose to public.  

I view this low priority work.  My opinion is the code analysis against only the all options enabled generated projects because it does add some time in build process.

Any reason why we wouldn't use our suppression file?

**Comment by ralarcon** [comment]
We keep in mind this to implement it when possible and check how impacts in the build (maybe we can only add stylecop checking to the projects which includes all the features, that is, 9 projects. 

We can use our suppression file, of course ;-)

**Comment by mrlacey** [comment]
We shouldn't use the same suppression file as we use in the tool code because:
- it includes exclusions which are currently only temporary and we will eventually (hopefully) remove
- it doesn't exclude things which don't apply to generated code. (e.g. file copyright headers)
- it doesn't correspond with [the guidelines](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) we claim to follow.
- there are issues we previously excluded but have now been addressed but still exist in the generated code.
- even with our current suppression set, there are inconsistencies that are highlighted in the generated code.

Rather than vague theorizing about this, I feel it just needs to be done.

**Comment by ralarcon** [comment]
Agree. We can setup the suppresion file we want. My understanding is that as time goes, we will be improving our supression to reflect all things you claim, so we can use the same quality bar for generated code and for our code. It is not possible to go 0 - 100 in one shot.

By the way, this is not theorization, it is more a solution aproach to be implemented in short term.

**Comment by mrlacey** [comment]
Have created a branch at https://github.com/Microsoft/WindowsTemplateStudio/tree/Issue461-StyleCopOfGenerated    
Will hold off merging with dev until after 1.1 is released.  
Points of note:
- Had to disable tests running in parallel for this to work (also have done this in the VB branch.) The generation code is very tightly coupled to some global static objects which make testing tricky. I've created #497 to track improving this.
- Added a new PostAction type (Search and Replace) will do multiple search&replace actions within a file. This was necessary to make the required modifications within the csproj as I came up against a limit of the current merge post action. I can imagine this potentially being used for other things in the future so might be good to have around anyway.
- Test sets warnings as errors to make sure that any issues are caught by the test.
- Suppression file that's used is [here](https://github.com/Microsoft/WindowsTemplateStudio/blob/Issue461-StyleCopOfGenerated/code/test/Templates.Test/TestData/StyleCop/GlobalSuppressions.cs). I've tried to suppress as little as possible and only what I could justify. Yes, it's very different to what we use in the generator code.
- To make the tests pass with the suppressions provided I had to modify quite a few of the existing template files. The plus side of this is that there is improved standardization and consistency in the generated code.
- Tests run on all project/framework combinations with all pages and features added. I strongly suspect that there will be some combinations of page&feature inclusion which don't pass the stylecop test due to missing or extra blank lines in the generated code, but think this is good enough.
- I've been having issues running all the tests with the Resharper TestRunner. Not jus these new tests and not an issue I want to investigate at the moment. Mentioned here as an FYI in case anyone else sees this. Running them with the VS runner doesn't have issues.

**Comment by ralarcon** [comment]
@mrlacey without having reviewed in detail all the previous comment just to mention that we are making deep changes related to the right click feature. 

Maybe you can hold and try to integrate to this branch when stabilized (by the way we expect to have a more-or-less stable stable branch by this week)... in other words, can this be hold to be integrated with the right-click work as we will have some collisions?

By the way, I think that your considerations are fine so we can try to have them in consideration and include those for the next release.

**Comment by mrlacey** [comment]
and it's in Woo Hoo!
